backend/src/common/services/calculation-progress/calculation-progress.service.ts

```diff
@@ -0,0 +1,190 @@
+import { type AchievementService } from '~/bundles/achievements/achievement.service.js';
+import { type ValueOf, Metric } from '~/bundles/achievements/enums/enums.js';
+import { type UserAchievementService } from '~/bundles/achievements/user-achievement.service.js';
+import { type GoalService } from '~/bundles/goals/goal.service.js';
+import { type NotificationService } from '~/bundles/notifications/notification.service.js';
+import { type WorkoutResponseDto } from '~/bundles/workouts/types/types.js';
+import { workoutService } from '~/bundles/workouts/workouts.js';
+import { COMPLETED_GOAL_VALUE } from '~/common/constants/constants.js';
+import { NotificationMessage } from '~/common/enums/enums.js';
+import {
+    calculateGoalProgress,
+    convertMetersToKilometers,
+} from '~/common/helpers/helpers.js';
+
+class CalculationProgressService {
+    private achievementService: AchievementService;
+    private goalService: GoalService;
+    private userAchievementsService: UserAchievementService;
+    private notificationService: NotificationService;
+
+    public constructor({
+        goalService,
+        achievementService,
+        userAchievementService,
+        notificationService,
+    }: {
+        goalService: GoalService;
+        achievementService: AchievementService;
+        userAchievementService: UserAchievementService;
+        notificationService: NotificationService;
+    }) {
+        this.achievementService = achievementService;
+        this.goalService = goalService;
+        this.userAchievementsService = userAchievementService;
+        this.notificationService = notificationService;
+    }
+
+    public async calculateProgress(
+        userId: number,
+        lastWorkout: WorkoutResponseDto,
+    ): Promise<void> {
+        await this.updateGoal(userId, lastWorkout);
+        await this.updateAchievement(userId, lastWorkout);
+    }
+
+    private async updateGoal(
+        userId: number,
+        lastWorkout: WorkoutResponseDto,
+    ): Promise<void> {
+        let hasCompletedGoal = false;
+
+        const { items: goals } = await this.goalService.findAll({
+            userId,
+        });
+        const { items: workouts } = await workoutService.findAll({
+            userId,
+        });
```

Comment:
you don't need to get all workouts, could already get pre-filtered for specific activity type, the same applies to goals/achievements etc.